### ISSUE TITLE
Method Geometry#contains? not defined.

### ISSUE THREAD

**Issue by TangMonk** [rgeo/rgeo#143]
I have following code;

```
p1 = factory.point 103.81749281429597, 30.724083064755405 
p2 = factory.point 103.8248780945283, 30.727447235175685 
p3 = factory.point 103.83126560723038, 30.716841883707545 
p4 = factory.point 103.82404639424398, 30.713477506095202 
p5 = factory.point 103.81749281429597, 30.724083064755405

linear_ring = factory.linear_ring([p1, p2, p3, p4])

linear_ring.contains? p1

>>RGeo::Error::UnsupportedOperation: Method Geometry#contains? not defined.
```

**Comment by TangMonk** [comment]
and I check the source code of `contains?`, it directly throw an error: https://github.com/rgeo/rgeo/blob/master/lib/rgeo/feature/geometry.rb#L357

Is is not implemented yet?

**Comment by brycejohnston** [comment]
I've been able to use contains?, overlaps?, and some of the other methods on geometries overridden in geometry.rb. The dependencies for GEOS, proj4 are installed and rgeo can find them? The implementation for contains is from GEOS itself:

https://github.com/rgeo/rgeo/blob/8eb105a96a7d40d9528d2d01f158974d07db597d/lib/rgeo/geos/zm_feature_methods.rb#L102

https://github.com/rgeo/rgeo/blob/8eb105a96a7d40d9528d2d01f158974d07db597d/ext/geos_c_impl/geometry.c#L603

**Comment by YenTheFirst** [comment]
What is the type of 'factory', and what is the type of 'linear_ring'?

For most planar types, #contains? is implemented by one of the FFI calls, and requires GEOS to be installed when installing the gem.
For the Geographic::Spherical types, #contains? is not implemented at all. If you want to use #contains?, you'll need to cast it to a different factory type.
For the Cartesian bounding box type, #contains? is implemented in ruby.

**Comment by TangMonk** [comment]
@YenTheFirst the type of factory is `Geographic`

**Comment by rohanthewiz** [comment]
You are probably missing libgeos-dev (or the equivalent for your operating system)
The README gives all you need: 

> Some features also require the following:

> GEOS 3.2 or later is highly recommended. (3.3.3 or later preferred.) Some functions will not be available without it. This C/C++ library may be available via your operating system's package manager (sudo aptitude install libgeos-dev for debian based Linux distributions, yum install geos geos-devel for redhat based Linux distributions), or you can download it from http://trac.osgeo.org/geos.

**Comment by rohanthewiz** [comment]
"Some features" (geometrical calculations etc) are what you need :-). For me it came down to:
`sudo apt-get install libgeos-dev`.
You can also verify all is good in your code at a pry console for example with `RGeo::Geos.supported? #=> true`.
If you use FFI (`RGeo::Geos.preferred_native_interface = :ffi`), then the following will also be true `RGeo::Geos.ffi_supported?`.

**Comment by ConorNugent** [comment]
I also was a bit surprised to run into issues with 'contains?' and posted the following question on stackoverflow about it

http://stackoverflow.com/questions/39675580/polygon-contains-and-other-operations-on-geometries-not-supported

If 'contains?' isn't supported I'm not sure how to query which objects are inside a given boundary. Maybe I should take a different approach?

**Comment by alibby** [comment]
@soynog  Try making sure you're using the same ruby and gemset when inside and outside of rails.  That'd be my guess as to the cause.  ```which ruby``` and ```gem environment```
when in the two different.

**Comment by soynog** [comment]
Thanks for the help, @alibby. Not quite sure if I'm understanding properly what you're saying I should do, but:

When I run ```which ruby``` in my terminal (zsh) I get:
```/Users/nicolasgarcia/.rbenv/shims/ruby```

whereas when I run it in the rails console I get:
```/Users/nicolasgarcia/.rbenv/versions/2.1.7/bin/ruby ```

Running ```gem environment``` in the terminal yields:
```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.2.5
  - RUBY VERSION: 2.1.7 (2015-08-18 patchlevel 400) [x86_64-darwin15.0]
  - INSTALLATION DIRECTORY: /Users/nicolasgarcia/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0
  - RUBY EXECUTABLE: /Users/nicolasgarcia/.rbenv/versions/2.1.7/bin/ruby
  - EXECUTABLE DIRECTORY: /Users/nicolasgarcia/.rbenv/versions/2.1.7/bin
  - SPEC CACHE DIRECTORY: /Users/nicolasgarcia/.gem/specs
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-darwin-15
  - GEM PATHS:
     - /Users/nicolasgarcia/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0
     - /Users/nicolasgarcia/.gem/ruby/2.1.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/nicolasgarcia/.rbenv/versions/2.1.7/bin
     - /usr/local/Cellar/rbenv/1.1.0/libexec
     - /Users/nicolasgarcia/.rbenv/shims
     - /Users/nicolasgarcia/.rbenv/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /opt/X11/bin
```

But the rails console doesn't know what I'm talking about (I think the command will only take gem names).

Is that helpful? I should be running 2.1.7 across the board.

**Comment by alibby** [comment]
@soynog yeah, I was not sure if I was being clear or not.  So perhaps going into irb and rails console and looking at the RUBY_VERSION.  Something like this:


```
alibby-2:~ alibby$ irb
2.3.1 :001 > RUBY_VERSION
 => "2.3.1"
2.3.1 :002 > quit
alibby-2:~ alibby$ cd development/zugunroute/zugunroute
alibby-2:zugunroute alibby$ rails c
Loading development environment (Rails 3.2.22)
irb: warn: can't alias context from irb_context.
2.1.5 :001 > RUBY_VERSION
 => "2.1.5"
2.1.5 :002 >
```

If they both yield 2.1.7, we're probably looking at different gemsets as the culprit.

When you ran gem environment above, were you in the rails project folder? 

The other thing to try to figure out is what ruby binary the shim points to.
When you typed which ruby via zsh above it referenced a shim.  If you go to that
folder on the filesystem and see where the shim points, we might be able to determine
if the two rubies being run are the same.

I'm not terribly familiar with rbenv so I'm feeling around a little in the dark.

Good luck.

Andy

**Comment by soynog** [comment]
Thanks again! Yes, RUBY_VERSION yields "2.1.7" both in rails and in irb. And I was indeed in my rails project folder when I ran gem environment.

I think you might be right about gemsets being the culprit--this problem started occurring when I tried to switch ruby version managers to rvm because I thought it might solve a problem I was having (it didn't). As far as I can tell rvm has gemset functionality built in, whereas rbenv does not. It's not obvious to me from inspecting the shim which ruby binary it's pointing to. The last two lines are:
```
export RBENV_ROOT="/Users/nicolasgarcia/.rbenv"
exec "/usr/local/Cellar/rbenv/1.1.0/libexec/rbenv" exec "$program" "$@"
```

I might try adding a gemset manager add-on to rbenv and see if I can figure something out that way.

This seems like we're maybe getting out of RGeo territory now, but I very much appreciate your help.
